File: FlutterProjects/Practice/Backend with JWT- v6 Upgrading/src/main/java/com/mrahmed/HRandPayrollManagementSystem/service/PaySlipService.java
package com.mrahmed.HRandPayrollManagementSystem.service;

import com.mrahmed.HRandPayrollManagementSystem.entity.PaySlip;
import com.mrahmed.HRandPayrollManagementSystem.entity.User;
import com.mrahmed.HRandPayrollManagementSystem.entity.Salary;
import com.mrahmed.HRandPayrollManagementSystem.repository.PaySlipRepository;
import com.mrahmed.HRandPayrollManagementSystem.repository.UserRepository;
import com.mrahmed.HRandPayrollManagementSystem.repository.SalaryRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
public class PaySlipService {

    @Autowired
    private PaySlipRepository paySlipRepository;

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private SalaryRepository salaryRepository;

    // Create PaySlip
    public PaySlip createPaySlip(Long employeeId, Long managerId, Long salaryId, double totalAmount, String paymentMethod) {
        User paidBy = userRepository.findById(managerId).orElseThrow(() -> new RuntimeException("Manager not found"));
        User receivedBy = userRepository.findById(employeeId).orElseThrow(() -> new RuntimeException("Employee not found"));
        Salary salary = salaryRepository.findById(salaryId).orElseThrow(() -> new RuntimeException("Salary not found"));

        PaySlip paySlip = new PaySlip();
        paySlip.setTotalAmount(totalAmount);
        paySlip.setBillingDate(LocalDateTime.now());  // Set the billing date to the current date
        paySlip.setPaymentMethod(paymentMethod);
        paySlip.setStatus("PAID");  // Set initial status as "PAID"
        paySlip.setPaidBy(paidBy);
        paySlip.setReceivedBy(receivedBy);
        paySlip.setSalary(salary);

        return paySlipRepository.save(paySlip);  // Save the pay slip to the database
    }

    // Get Payslips by Employee ID
    public List<PaySlip> getPayslipsByEmployee(Long employeeId) {
        return paySlipRepository.findByReceivedById(employeeId);
    }

    // Get Payslips by Manager ID
    public List<PaySlip> getPayslipsByManager(Long managerId) {
        return paySlipRepository.findByPaidById(managerId);
    }

    // Get Payslips by Status
    public List<PaySlip> getPayslipsByStatus(String status) {
        return paySlipRepository.findByStatus(status);
    }

    // Get Payslips by Billing Date Range
    public List<PaySlip> findByBillingDateBetween(LocalDateTime startDate, LocalDateTime endDate) {
        return paySlipRepository.findByBillingDateBetween(startDate, endDate);
    }

    // Get Payslip by Salary ID
    public PaySlip getPayslipBySalaryId(Long salaryId) {
        return paySlipRepository.findBySalaryId(salaryId).orElseThrow(() -> new RuntimeException("Payslip not found"));
    }

    // Get Payslips by Payment Method (e.g., Cash, Bank)
    public List<PaySlip> getPayslipsByPaymentMethod(String paymentMethod) {
        return paySlipRepository.findByPaymentMethod(paymentMethod);
    }

    // findByReceivedByIdAndStatus
    public List<PaySlip> getPayslipsByEmployeeAndStatus(Long employeeId, String status) {
        return paySlipRepository.findByReceivedByIdAndStatus(employeeId, status);
    }


    // Count Payslips by Status
    public Long countPayslipsByStatus(String status) {
        return paySlipRepository.countByStatus(status);
    }
}

File: FlutterProjects/Practice/hr_and_pms - v4 Upgrading/lib/features/paySlip/service/PaySlipService.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class PaySlipService {
  final String baseUrl = 'http://localhost:8080/api/payslips';

  // Fetch PaySlips by Employee ID
  Future<List<dynamic>> getPayslipsByEmployee(int employeeId) async {
    final response = await http.get(Uri.parse('$baseUrl/$employeeId'));

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load payslips for employee');
    }
  }

  // Fetch PaySlips by Manager ID
  Future<List<dynamic>> getPayslipsByManager(int managerId) async {
    final response = await http.get(Uri.parse('$baseUrl/manager/$managerId'));

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load payslips for manager');
    }
  }

  // Fetch PaySlips by Status
  Future<List<dynamic>> getPayslipsByStatus(String status) async {
    final response = await http.get(Uri.parse('$baseUrl/status/$status'));

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load payslips by status');
    }
  }

  // Fetch PaySlips by Employee ID and Status
  Future<List<dynamic>> getPayslipsByEmployeeAndStatus(
      int employeeId, String status) async {
    final response =
    await http.get(Uri.parse('$baseUrl/$employeeId/status/$status'));

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception(
          'Failed to load payslips for employee $employeeId with status $status');
    }
  }

  // Fetch PaySlips by Billing Date Range
  Future<List<dynamic>> getPayslipsByDateRange(
      String startDate, String endDate) async {
    final response = await http
        .get(Uri.parse('$baseUrl/date-range?startDate=$startDate&endDate=$endDate'));

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load payslips by date range');
    }
  }

  // Fetch a PaySlip by Salary ID
  Future<dynamic> getPayslipBySalaryId(int salaryId) async {
    final response = await http.get(Uri.parse('$baseUrl/salary/$salaryId'));

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load payslip for salary ID $salaryId');
    }
  }

  // Fetch PaySlips by Payment Method
  Future<List<dynamic>> getPayslipsByPaymentMethod(String paymentMethod) async {
    final response =
    await http.get(Uri.parse('$baseUrl/payment-method/$paymentMethod'));

    if (response.statusCode == 200) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to load payslips by payment method');
    }
  }

  // Count PaySlips by Status
  Future<int> countPayslipsByStatus(String status) async {
    final response = await http.get(Uri.parse('$baseUrl/count/status/$status'));

    if (response.statusCode == 200) {
      return json.decode(response.body) as int;
    } else {
      throw Exception('Failed to count payslips by status');
    }
  }

  // Create PaySlip
  Future<dynamic> createPaySlip({
    required int employeeId,
    required int managerId,
    required int salaryId,
    required double totalAmount,
    required String paymentMethod,
  }) async {
    final response = await http.post(
      Uri.parse('$baseUrl/create'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode({
        'employeeId': employeeId,
        'managerId': managerId,
        'salaryId': salaryId,
        'totalAmount': totalAmount,
        'paymentMethod': paymentMethod,
      }),
    );

    if (response.statusCode == 200 || response.statusCode == 201) {
      return json.decode(response.body);
    } else {
      throw Exception('Failed to create payslip');
    }
  }
}
